daemons/attrd/attrd_attributes.c

```diff
@@ -14,11 +14,11 @@
 #include <stdlib.h>
 #include <glib.h>
 
-#include <crm/msg_xml.h>
 #include <crm/common/logging.h>
 #include <crm/common/results.h>
 #include <crm/common/strings_internal.h>
 #include <crm/common/xml.h>
+#include <crm/common/xml_names.h>
```

Comment:
I thought you didn't like transitive includes. Maybe since these are so closely related. Works for me.

Reply:
I'm going to get confirmation on that before I go through and change these... we can even take the stance "never include `xml_names.h` or `xml_names_internal.h` directly" if we want.

A lot of these files previously included only `msg_xml.h` and not `xml.h`. I get the feeling they were relying on THAT transitive include -- `msg_xml.h` and `xml.h` previously included each other circularly.

Reply:
Transitive includes are a gray area. It's more of a concern in headers than in source files, to guarantee that the header is always usable without the source files having to know to include random other stuff they might not directly need. Source files aren't as big a deal since only the one file can be affected. I'd still prefer to avoid unrelated transitive includes (e.g. not relying on some random header to include stdbool.h), but even then it's not horrible. For related includes between internal headers, it's not important.

Eventually I'd like to get Pacemaker to where we have a single public header and single internal header per library (much like glib has glib.h to include everything). Then we can more consistently avoid other transitive includes without too much mess.

Reply:
> I'm going to get confirmation on that before I go through and change these... we can even take the stance "never include `xml_names.h` or `xml_names_internal.h` directly; instead always include `xml.h`" if we want.

For convenience, I have followed that approach when separating one header from another, since so many files already include the original header. It's just a quick hack along the way to the more defined header-per-library approach.

> A lot of these files previously included only `msg_xml.h` and not `xml.h`. I get the feeling they were relying on THAT transitive include -- `msg_xml.h` and `xml.h` previously included each other circularly.

Yeah I fought circular includes for a long time. xml.h never included msg_xml.h though AFAIK.

Reply:
Personally, I like it when a header file includes everything it needs for itself - if the header uses uint32_t, it should include stdint.h.  If it uses GList, it should include glib.h.  That way it's not on the user of the header file to figure it out.  I am also fine with the one giant header that includes all the others so I have to remember less.

Reply:
> xml.h never included msg_xml.h though AFAIK.

In the main branch, they each include each other.

Reply:
Ah, I missed that when it was recently added (94a1afce)